File: leetcode-python/String/LC0125-验证回文串.py
'''
lc 验证回文串

输入: "A man, a plan, a canal: Panama"
输出: true
输入: "race a car"
输出: false
'''

class Solution:
    def isPalindrome(self, s: str) -> bool:
        b = [i for i in s.lower() if i.isalnum()]
        return b == b[::-1]

class Solution:
    def isPalindrome(self, s: str) -> bool:
        res = str.lower(re.sub('[\W_+]','',s))
        return res == res[::-1]

class Solution:
    def isPalindrome(self, s: str) -> bool:
        '''
        b = ''
        for i in s.lower():
            if i.isalnum():
                b += i
        '''

        b = [i for i in s.lower() if i.isalnum()]
        ans = ''.join(b)

        l, r = 0, len(ans) - 1
        while l < r:
            if ans[l] != ans[r]:
                return False
            l += 1
            r -= 1
        return True


        '''
        for i in range(len(b) / 2):
            if b[i] != b[len(b) - i]:
                return False
        return True
        '''
    '''
    执行用时44 ms, 在所有 Python3 提交中击败了96.42%的用户
    '''

File: leetcode-go/String/LC0125-验证回文串.go
package main

import "strings"

/**
 * 验证回文串
 * 给定一个字符串，验证它是否是回文串，只考虑字母和数字字符，可以忽略字母的大小写。
 * 输入: s = "A man, a plan, a canal: Panama"
 * 输出：true
 * 解释："amanaplanacanalpanama" 是回文串。
 * 难度: 简单
 */


func isPalindrome(s string) bool {
	str := ""
	for _, v := range s {
		// 只考虑字母和数字字符
		if unicode.IsLetter(v) || unicode.IsDigit(v) {
			// 忽略字母的大小写
			str += strings.ToLower(string(v))
		}
	}
	// 验证它是否是回文串
	return isReverseString(str)
}

func isReverseString(s string) bool {
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		if s[i] != s[j] {
			return false
		}
	}
	return true
}

func isReverseString2(s string) bool {
	for i := 0; i < len(s)/2; i++ {
		if s[i] != s[len(s)-1-i] {
			return false
		}
	}
	return true
}
